.cursor/rules/rust-coding-style.mdc

```diff
@@ -0,0 +1,112 @@
+---
+description: Rust Coding Style
+globs: "**/*.rs"
+alwaysApply: false
+---
+# Rust Coding Style
+
+## Project-Specific Patterns
+
+- Use the 2024 edition of Rust
+- Prefer `derive_more` over manual trait implementations
+- Feature flags in this codebase use the `#[cfg(feature = "...")]` pattern
+- Invoke `cargo clippy` with `--all-features` and `--all-targets` from the root
+- Use `rustfmt` to format the code
+- Use `#[expect(lint, reason = "...")]` over `#[allow(lint)]`
+
+## Type System
+
+- Create strong types with newtype patterns for domain entities
+- Implement `hash_graph_types` traits for custom domain types
+- Consider visibility carefully (avoid unnecessary `pub`)
+
+```rust
+#[derive(Debug, Copy, Clone, Eq, Hash, PartialEq, derive_more::Display)]
+pub struct UserId(Uuid);
+```
+
+## Async Patterns
+
+- Use `impl Future<Output = T> + Send` in trait definitions
+- When returning futures from trait methods, use the async_trait pattern:
+
+```rust
+fn get_data(
```

Comment:
this isn't required is it? I'd say the opposite actually - if you trait that returns -> impl Future + Send first try to implement it as async fn and only desugar if you can't.